internal/namespace/reachabilitygraphbuilder.go

```diff
@@ -0,0 +1,226 @@
+package namespace
+
+import (
+	"fmt"
+
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+type reachabilityOption int
+
+const (
+	reachabilityFull reachabilityOption = iota
+	reachabilityOptimized
+)
+
+func computeReachability(ts *NamespaceTypeSystem, relationName string, option reachabilityOption) (*core.ReachabilityGraph, error) {
+	targetRelation, ok := ts.relationMap[relationName]
+	if !ok {
+		return nil, fmt.Errorf("relation `%s` not found under type `%s` missing when computing reachability", relationName, ts.nsDef.Name)
+	}
+
+	if !ts.HasTypeInformation(relationName) && !ts.IsPermission(relationName) {
+		return nil, fmt.Errorf("relation `%s` missing type information when computing reachability", relationName)
+	}
+
+	graph := &core.ReachabilityGraph{
+		EntrypointsBySubjectType:     map[string]*core.ReachabilityEntrypoints{},
+		EntrypointsBySubjectRelation: map[string]*core.ReachabilityEntrypoints{},
+	}
+
+	usersetRewrite := targetRelation.GetUsersetRewrite()
+	if usersetRewrite != nil {
+		return graph, computeRewriteReachability(graph, usersetRewrite, core.ReachabilityEntrypoint_DIRECT_OPERATION_RESULT, targetRelation, ts, option)
```

Comment:
I'm confused why we would be passing in `core.ReachabilityEntrypoint_DIRECT_OPERATION_RESULT` here when the whole point of `computeRewriteReachability` is to determine what type of reachability this rewrite has.

Reply:
Because until we see an intersection or exclusion, it is assumed to be a direct result. Only once we've seen one (or more) of those operations, do we need to make the entire thing conditional